rawfile/filesystem/base.py

```diff
@@ -0,0 +1,82 @@
+from abc import ABC, ABCMeta, abstractmethod
+
+
+class NotSupportedError(NotImplementedError):
+    """
+    Exception raised when a feature is not supported by the current filesystem.
+    """
+
+    def __init__(self, feature: str) -> None:
+        super().__init__(
+            f"Feature '{feature}' is not supported by the current filesystem."
+        )
+
+
+class FileSystem(ABC, metaclass=ABCMeta):
+    def __init__(self, device: str, mountpoint: str | None = None):
+        """
+        Initialize the file system with the given device.
+        """
+        self.device = device
+        self.mountpoint = mountpoint
+
+    @classmethod
+    def __subclasshook__(cls, subclass):
+        """
+        Check if the subclass implements the required methods.
+        """
+        required_methods = [
+            "format_fs",
+            "mount",
+            "unmount",
+            "resize",
+        ]
+        for method in required_methods:
+            if not (hasattr(subclass, method) and callable(getattr(subclass, method))):
+                return False
+        return True
+
+    @abstractmethod
+    def create_snapshot(self):
+        """
+        Create a snapshot of the current state of the file system.
+        """
+        raise NotSupportedError("create_snapshot")
+
+    @abstractmethod
+    def delete_snapshot(self):
+        """
+        Delete the most recent snapshot of the file system.
+        """
+        raise NotSupportedError("delete_snapshot")
+
+    @abstractmethod
+    def format_fs(self, options: list[str] = []):
+        """
+        Format the file system.
+        """
+        raise NotSupportedError("format_fs")
+
+    @abstractmethod
+    def mount(self, mountpoint: str, options: list[str] = []):
+        """
+        Mount the file system.
+
+        Should also do
+            self.mountpoint = mountpoint
```

Comment:
but here the mountpoint in args != self.mountpoint right?

Reply:
Yeah, Also self.mountpoint is could be None